src/lib/Sympa/Request/Handler/add_list_admin.pm

```diff
@@ -0,0 +1,198 @@
+    my $robot   = $list->{'domain'};
+    my $role = $request->{role};
+    my $sender = $request->{sender};
+    my $updater_user = $sender || Sympa::get_address($list, 'listmaster');
+    my $user;
+    $user->{email} = Sympa::Tools::Text::canonic_email($request->{email});
+    $user->{visibility} = $request->{visibility};
+    $user->{profile} = $request->{profile};
+    $user->{reception} = $request->{reception};
+    $user->{gecos} = $request->{gecos};
+    $user->{info} = $request->{info};
+
+    unless ($list) {
+        $log->syslog('err', 'Trying to add list admin without list object.');
+        $self->add_stash($request, 'intern');
+        return undef;
+    }
+    
+    my %mandatory_parameters = (
+        email => $user->{email},
+        role  => $role,
+    );
+    
+    my @missing_parameters;
+    foreach my $mandatory_param (keys %mandatory_parameters) {
+        unless ($mandatory_parameters{$mandatory_param}) {
+            push @missing_parameters, $mandatory_param;
+        }
+    }
+    
+    if( scalar @missing_parameters) {
+        $log->syslog('err', 'Trying to add list admin without giving %s.', join (', ', @missing_parameters));
+        $self->add_stash(
+            $request, 'user', 'missing_parameters',
+            {p_name => join ', ', @missing_parameters}
+        );
+        return undef;
+    }
+    
+    unless ($role eq 'owner' or $role eq 'editor') {
+        $self->add_stash(
+            $request, 'user', 'syntax_errors',
+            {p_name => 'role'}
+        );
+        $log->syslog('err', 'Error while adding %s as %s to list %s@%s. Bad parameter(s): %s',
+            $user->{email}, $role, $listname, $robot, 'role');
+        return undef;
+    }
+    
+    my $schema = $Sympa::ListDef::user_info{$role};
+    my @param_in_error;
+    foreach my $param (keys %{$schema->{format}}) {
+        if ($user->{$param}) {
+            unless($user->{$param} =~ /$schema->{format}{$param}{file_format}/) {
+                push @param_in_error, $param;
+            }
+        }
+    }
+     if (scalar @param_in_error) {
+        $self->add_stash(
+            $request, 'user', 'syntax_errors',
+            {p_name => join ', ', @param_in_error}
+        );
+        $log->syslog('err', 'Error while adding %s as %s to list %s@%s. Bad parameter(s): %s',
+            $user->{email}, $role, $listname, $robot, join ', ', @param_in_error);
+        return undef;
+     }
+    # Check if user is already admin of the list.
+    if ($list->is_admin($role, $user->{email})) {
+        $self->add_stash(
+            $request, 'user',
+            'already_list_admin',
+            {email => $user->{email}, role => $role, listname => $list->{'name'}}
+        );
+        $log->syslog('err', 'User "%s" has the role "%s" in list "%s@%s" already',
+            $user->{email}, $role, $listname, $robot);
+        return undef;
+    } else {
+        unless ($list->add_list_admin($role, $user)) {
```

Comment:
Though I haven't investigated, how this will work when `owner_domain` feature is enabled?

Reply:
That's an interesting question.
I'll check it and maybe add a test to t/RequestHandler_add_list_admin.t.

Reply:
OK. Currently, it does not react to this parameter.
Currently, it looks like `owner_domain` is used in wwsympa.fcgi only, which makes sense because it is the only way for end users to add list owners.
However, shouldn't it be used in List.pm directly, anytime we add or remove a list admin? Maybe I'm misunderstanding how this parameter works, but is would make sense to enforce such constraints in the central place that is List.pm.
In that case, that should be addressed in another PR. Otherwise, some code similar to wwsympa.fcgi should be added to the request handler. But that looks like code duplication to me.

Reply:
I opened the issue #597 to discuss this topic.